src/EditorFeatures/Core/InheritanceMargin/AbstractInheritanceMarginService.cs

```diff
@@ -0,0 +1,159 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.FindSymbols.FindReferences;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.SymbolMapping;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.InheritanceMargin
+{
+    internal abstract partial class AbstractInheritanceMarginService : IInheritanceMarginService
+    {
+        public async Task<ImmutableArray<InheritanceMemberItem>> GetInheritanceInfoAsync(
+            Document document,
+            CancellationToken cancellationToken)
+        {
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var featureEnabled = options.GetOption(InheritanceMarginOptions.ShowInheritanceMargin);
+            if (!featureEnabled)
+            {
+                return ImmutableArray<InheritanceMemberItem>.Empty;
+            }
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var allDeclarationNodes = GetMembers(root);
+            if (allDeclarationNodes.IsEmpty)
+            {
+                return ImmutableArray<InheritanceMemberItem>.Empty;
+            }
+
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var mappingService = document.Project.Solution.Workspace.Services.GetRequiredService<ISymbolMappingService>();
+            using var _ = ArrayBuilder<InheritanceMemberItem>.GetInstance(out var builder);
+
+            // Iterate all the members symbol to find each their inheritance chain information
+            foreach (var memberDeclarationNode in allDeclarationNodes)
+            {
+                var member = semanticModel.GetDeclaredSymbol(memberDeclarationNode, cancellationToken);
+                if (member != null && !member.IsStatic && !member.IsErrorType())
```

Comment:
how do you get IsErrorType to ever be true here?

Reply:
I dont' think it would be error type, and I think I add this for sake fo safety, removed.